Artifact offhand ilvl fix

dejoblue opened the pull request:
Artifacts were still now showing the correct ilvl on the offhands, even
with effective level. If you link the Artifact from the offhand it shows
the level as 750, even using item link, so effective level doesn;t
work...

dejoblue committed:
Artifact Offhand Fix

Artifacts were still now showing the correct ilvl on the offhands, even
with effective level. If you link the Artifact from the offhand it shows
the level as 750, even using item link, so effective level doesn;t
work...

Kakjens commented:
Changed base for both pulls.
Don't have Legion expansion, so can't test.
I think the code is missing verification that:
1 the artifact is equipped in main hand;
2. it's artifact for correct spec.
And as I previously said, there might be cases when the artifact for which GetDetailedItemLevelInfo returns the right ilvl is offhand instead of mainhand.

dejoblue commented:
Artifacts only equip to the main hand and you can only equip an artifact for the spec you are in. If you are in a spec and change it is automatically swapped out by the default UI.

1. We don't need a check for low level characters because it will not be artifact and will do the normal thing.

2. Artifacts only equip to the main hand and you can only equip an artifact for the spec you are in. If you are in a spec and change it is automatically swapped out by the default UI.

Yeah, the app must be messing up the bases I use for some reason.

Kakjens commented:
1. What did I mean was it should be possible to change (or un-equip) the main-hand weapon. For example, buying one dagger from vendor and equipping it.
2. I think I saw a screenshot where artifact for wrong spec was equipped. Might be my memory deceiving me or some bug.
Looking at protection paladin artefacts [Truthguard](http://www.wowhead.com/item=128866) and [Oathseeker](http://www.wowhead.com/item=128867), it seems like the main artefact in that case is in off-hand.
Thinking now, wouldn't it be more correct to calculate artifact ilvl as linear function from GetAveragreitemLevel and ilvl of currently equipped items which are not weapons?

dejoblue commented:
You aren't understanding how Artifact weapons work. You bring up good points but Artifacts do not work that way and regular items are fully independent.

1. You can do this already, without my change. The main hand and off hand are completely independent UNLESS you have an artifact weapon. The Artifacts are one item that you equip in the main hand. then it shows an icon in the offhand for dual wield or sword and board specs. You cannot unequip one or the other, only both. It is one item that occupies one item slot in your bags, but that shows two items on the paper doll. And as I stated, the main hand LINK, when you link it in chat, what we use to get effective ilvl is always 750 for the offhand and the effective ilvl for the main hand. local effectiveLevel = GetDetailedItemLevelInfo(**itemLink**) 

2. The only way this happens is if you do not have an artifact for the spec you change too. At that point the equipped artifact slots, both main hand and offhand, are turned red and not usable, you are unarmed. However, DCS still shows their ilvl just fine.

> Thinking now, wouldn't it be more correct to calculate artifact ilvl as linear function from GetAveragreitemLevel and ilvl of currently equipped items which are not weapons?

I think it is too granular. Artifacts gain ilvl from 3 different gems they have  which additively increase the overall ilvl of the Artifact. So a 1 ilvl change may not be reflected. As well, there is a direct call we are already using. Running a bunch of math to find it is inefficient and possibly inaccurate.

Kakjens commented:
I really don't know how artifact weapons work. Since you do have experience with them will trust you. But seeing a screenshot of how it looks on protection paladin would convince me.
Actually GetAveragreItemLevel  returns floating point numbers, so even small increases would get caught. Other thing, if artifact weapon can only be fully equipped/un-equipped then there's indeed no need for crazy (and probably complex) calculations.

dejoblue commented:
You appear to actually be right and I am the one that doesn't understand artifacts! I am making a video to show my findings.

dejoblue commented:
Here is a video to show the bugs and explain artifacts. Sorry it is 10 minutes long lol. Audio is only in left side for some reason...
https://www.youtube.com/watch?v=i9FE1ElrXNw

dejoblue commented:
I'll work on a fix with this method to use the highest of the main hand or offhand if it is an Artifact weapon. This will get us set for the time being. It appears that other than artifacts the only gear with ilvl issues is old Timewarped gear, from previous expansions, so that is likely a small number of players. 

Checking on an alt, old Warforged gear also has this issue.

Kakjens commented:
Thanks for video. On my left side the headphones recently became not properly working - the sound is more silent. Nevertheless, managed to hear by cranking volume up.
Issue with artifact weapons indeed is easy to fix - the max of both numbers. Can't quite imagine the most efficient way to do it - there are several checks that need to be performed.
Issue with old gear can be solved by writing a bug report to Blizzard. I agree that the affected player-base should be small for them to be ignorable.
I feel that the ilvl would be interesting in timewalking dungeons.
By the way, how did you catch that prot warrior has the issue? If my memory is right, certain warlock spec should also have it.

dejoblue commented:
> By the way, how did you catch that prot warrior has the issue? If my memory is right, certain warlock spec should also have it.

I looked at my prot paladin, your links to truthguard and oathseeker made me curious. Then in game the shield had full text but the sword did not and right clicking to open the artifact panel opened the shield.

I looked on the AH to buy a relic to level the artifact but there weren't any. Then I thought of my warrior. Did the same, no relics on the AH. then I noticed that I had almost finished the first quests in Stormheim and each zone gives a relic when you finish it, so I did like 5 quests, and got a relic and the shield went to 753 and the sword did not. Confirmed!

THEN I noticed the gear on my warrior was waaay wrong. Instead of showing the base it showed way too high...

I put n a bug report, https://us.battle.net/forums/en/wow/topic/20759375939

Kakjens commented:
To clarify: I expect for GetDetailedItemLevelInfo to return the ilvl what is in tooltip during mouse-over.

dejoblue commented:
That is what I meant, effectiveLevel is the variable, the issue must be with GetDetailedItemLevelInfo or GetInventoryItemLink, I edited the bug report to reflect that.

dejoblue commented:
New problem, Artifacts upgraded with the Netherlight Crucible do not show the proper ilvl...

https://i.imgur.com/wQgG1TR.jpg

We'll have to do tooltip string searches I guess...

Kakjens commented:
That sounds troubling. It might be possible that even tooltip parsing might [not work](https://us.battle.net/forums/en/wow/topic/20759165718).

dejoblue commented:
I tested scanning the tooltip last night but the item link is always wrong and returns what we already have. i posted on that thread. I hope we aren't supposed to use bonus IDs and or tables of each artifact or even upgrade-able gear everywhere.

Kakjens commented:
I feel that we can't do much until ilvl [stabilizes](https://us.battle.net/forums/en/wow/topic/20759316177).

dejoblue commented:
Yeah, but the tooltip is always right, there has to be a way to get it off the tooltip. I am gonna look for more methods. Here is one I will try tonight. 
https://us.battle.net/forums/en/wow/topic/14881058187

If you have time and can look for something that doesn't involve using itemlink to make a tooltip let me know if you find anything. I'll keep looking tonight and trying things and thinking about solutions. 

We will probably have to do this for bonus XP as well so I think it is worth figuring out even if we have to use itemlink because whatever method we use can be used for other info and wrapped into one scan for efficiency sake.

Anyway, once we have a solution, or decide there is no solution, we should push what we have to a final beta for a few days or a week, id like to see 2000+ beta users for the final beta 105 before we release.

Kakjens commented:
Seeing the code how did you try to parse the tooltip would help. I remember there was a code written by Phanx how s/he would implement it.
I think there are quite a few addons that rely on GetDetailedLevelInfo for ilvl. Insufficient testing for breaking the old stuff by Blizzard is nothing new.
Indeed, figuring out a reliable way to get stuff (in this particular case, ilvl) is the purpose of this pull.

dejoblue commented:
> Seeing the code how did you try to parse the tooltip would help.

The method I linked above. A couple of others, but hey all involve itemlink.

This is what I was last playing with:

```
local function DCS_Item_Level_Center()
	for _, v in ipairs(DCSITEM_SLOT_FRAMES) do
		local itemLink = GetInventoryItemLink("player", v:GetID())
		if not itemLink then
			v.ilevel:SetFormattedText("")
		else
			local _, _, itemRarity = GetItemInfo(itemLink)
			local effectiveLevel = GetDetailedItemLevelInfo(itemLink)
			local r, g, b = GetItemQualityColor(itemRarity)
			--print(itemLink, itemLevel)
			local tip = myTooltipFromTemplate or CreateFrame("GAMETOOLTIP", "myTooltipFromTemplate",nil,"GameTooltipTemplate")
			tip:SetOwner(WorldFrame, "ANCHOR_NONE")
			--local link = select(2, GetItemInfo(GetContainerItemLink(v)))
			tip:SetHyperlink(itemLink)
			local lines = tip:NumLines()
			--print(lines, text, itemLink)
			for i=1,lines do
			local text = _G["myTooltipFromTemplateTextLeft"..i]:GetText()
			--print(text)
				local startPos, endPos, firstWord, restOfString = string.find(text, "Item Level " .. "(%w+)[%s%p]*(.*)");
				--print(firstWord)
				if (firstWord ~= nil) then
					v.ilevel:SetText(firstWord)
					--print(firstWord)
				end
			end
			v.ilevel:SetTextColor(r, g, b)
		end
	end
end
```

Phanx is awesome, I think Phanx is a she, if you could find her method it would rock.

dejoblue commented:
This specifically:

```
local tip = myTooltipFromTemplate or CreateFrame("GAMETOOLTIP", "myTooltipFromTemplate",nil,"GameTooltipTemplate")
			tip:SetOwner(WorldFrame, "ANCHOR_NONE")
			--local link = select(2, GetItemInfo(GetContainerItemLink(v)))
			tip:SetHyperlink(itemLink)
			local lines = tip:NumLines()
			--print(lines, text, itemLink)
			for i=1,lines do
			local text = _G["myTooltipFromTemplateTextLeft"..i]:GetText()
			--print(text)
				local startPos, endPos, firstWord, restOfString = string.find(text, "Item Level " .. "(%w+)[%s%p]*(.*)");
				--print(firstWord)
				if (firstWord ~= nil) then
					v.ilevel:SetText(firstWord)
					--print(firstWord)
				end
			end
```

Kakjens commented:
Leaving aside possible access to global myTooltipFromTemplate (did you mean _G["myTooltipFromTemplate"]?) I don't see any obvious errors.
Below is the alternative version which I found:

```
local ITEM_LEVEL_PATTERN = ITEM_LEVEL:gsub("%%d", "(%%d+)")
local tooltip = CreateFrame("GameTooltip", "iLevelScanTooltip", nil, "GameTooltipTemplate")
tooltip:SetOwner(UIParent, "ANCHOR_NONE")
 
local function GetItemLevel(itemLink)
    if not itemLink then return end
 
    tooltip:ClearLines()
    tooltip:SetHyperlink(itemLink)
 
    for i = 2, tooltip:NumLines() do
        local text = _G["iLevelScanTooltipTextLeft"..i]:GetText()
 
        if(text and text ~= "") then
            local value = tonumber(text:match(ITEM_LEVEL_PATTERN))
 
            if value then
                return value
            end
        end
    end
end

```
That should also work for non-English clients.
Can't test on artifacts/MoP/WoD items but can on battleground bag and heirlooms.

dejoblue commented:
Same result. itemLink is borked ;/

Here is my integration of your example:

```
local function DCS_Item_Level_Center()
	for _, v in ipairs(DCSITEM_SLOT_FRAMES) do
		local ITEM_LEVEL_PATTERN = ITEM_LEVEL:gsub("%%d", "(%%d+)")
		local tooltip = CreateFrame("GameTooltip", "iLevelScanTooltip", nil, "GameTooltipTemplate")
			tooltip:SetOwner(UIParent, "ANCHOR_NONE")
		local itemLink = GetInventoryItemLink("player", v:GetID())
		if not itemLink then
			v.ilevel:SetFormattedText("")
		else
			tooltip:ClearLines()
			tooltip:SetHyperlink(itemLink)
			for i = 2, tooltip:NumLines() do
				local text = _G["iLevelScanTooltipTextLeft"..i]:GetText()
		 
				if(text and text ~= "") then
					local value = tonumber(text:match(ITEM_LEVEL_PATTERN))
		 
					if value then
							v.ilevel:SetText(value)
					end
				end
			end
			local _, _, itemRarity = GetItemInfo(itemLink)
			local r, g, b = GetItemQualityColor(itemRarity)
			v.ilevel:SetTextColor(r, g, b)
		end
	end
end
```

dejoblue commented:
Okay so that method, and the method I linked earlier for scanning tooltips, works for timewarped/warforged gear. The original code of this pull still does not.

Artifacts do not work with any method. We will have to do a greatest of ilvl comparison to get it to work. i like the method you posted better, it's cleaner, so I am going to work that up with a comparison and have it up tonight, hopefully.

dejoblue commented:
Here is the nearly complete working code. heirlooms, Warforged, old Warforged, Timewarped, old Timewarped and non netherlight upgraded artifacts are all displaying properly. Netherlight Crucible upgrades are still not showing and the tooltip scrape does not get the increased ilvl. We will have to check if their relics have at least level 1. Each relic's first level gives 5 item levels, so it will at least be simply +5 for each relic at level 12 added to the artifact's ilvl.

Anyway, here is the code. Anything you can think of to reduce it or make it cleaner?

```
gdbprivate.gdbdefaults.gdbdefaults.dejacharacterstatsShowItemLevelChecked = {
	ShowItemLevelSetChecked = true,
	DCSMainHandSlot = 0,
	DCSSecondaryHandSlot= 0,
}	

local function DCS_Item_Level_Center()
	local DCSMainHandSlot = gdbprivate.gdb.gdbdefaults.dejacharacterstatsShowItemLevelChecked.DCSMainHandSlot
	local DCSSecondaryHandSlot = gdbprivate.gdb.gdbdefaults.dejacharacterstatsShowItemLevelChecked.DCSSecondaryHandSlot
	for _, v in ipairs(DCSITEM_SLOT_FRAMES) do
		local itemLink = GetInventoryItemLink("player", v:GetID())
		if not itemLink then
			v.ilevel:SetFormattedText("")
		else
			local ITEM_LEVEL_PATTERN = ITEM_LEVEL:gsub("%%d", "(%%d+)")
			local tooltip = CreateFrame("GameTooltip", "iLevelScanTooltip", nil, "GameTooltipTemplate")
				tooltip:SetOwner(UIParent, "ANCHOR_NONE")
				tooltip:ClearLines()
				tooltip:SetHyperlink(itemLink)
			for i = 2, tooltip:NumLines() do
				local text = _G["iLevelScanTooltipTextLeft"..i]:GetText()
				if(text and text ~= "") then
					local value = tonumber(text:match(ITEM_LEVEL_PATTERN))
					if value then
						local _, _, itemRarity = GetItemInfo(itemLink)
						local r, g, b = GetItemQualityColor(itemRarity)
							v.ilevel:SetTextColor(r, g, b)
						if (itemRarity == 6) then
							if (v == CharacterMainHandSlot) then 
								DCSMainHandSlot = value
							elseif (v == CharacterSecondaryHandSlot) then 
								DCSSecondaryHandSlot = value
							end
							if (DCSMainHandSlot >= DCSSecondaryHandSlot) then
								v.ilevel:SetText(DCSMainHandSlot)
							else
								v.ilevel:SetText(DCSSecondaryHandSlot)
							end
						else
							v.ilevel:SetText(value)
						end
					end
				end
			end
		end
	end
end
```

Kakjens commented:
Good news on old gear (not like many will see it).
The Crucible stuff leaving aside, indeed, it seems the code needs changing. There are several things I don't quite understand. Making a branch with code after my changes.
And [here](https://github.com/dejoblue/DejaCharacterStats/tree/ilvl_scantooltip) it is.
Edit. So you say that Blizzard seems to be [started ](https://us.battle.net/forums/en/wow/topic/20759165718)fixing GetDetailedItemLevelInfo for old stuff?

dejoblue commented:
Yeah, I am pretty sure the code I used, where I say Phanx is awesome, was not returning the proper ilvl on warforged, titanforged, timewarped, etc. Now it is. Or I didn't test right or something.

Regardless, it is working now for everything, even artifacts, EXCEPT for artifacts upgraded through the netherlight crucible. I will make another short video showing the netherlight crucible to you.

BUT those artifacts upgraded through the nntherlight crucible DO have the right tooltip.

dejoblue commented:
Here is the video: https://youtu.be/SrAPrMRBQUM

Here is the global Artifact UI API I see everyone using: https://wow.gamepedia.com/API_C_ArtifactUI.GetRelicInfoByItemID

A blue post said they don;t have an API for the relics.

dejoblue commented:
Your tootlip branch works the same as my two versions and your original version. Artifacts do not show upgraded ilvl: https://i.imgur.com/lQTeDgF.jpg while everything else shows fine.

There is not a need to separate the weapons out with different tables. DCSITEM_SLOT_FRAMES works perfectly fine.

There is no need to know a character's spec. All we are doing is looking at the items.

Other than separation, and a lot more code, you did what I already did.

I am going to commit changes to the pull so you see what I have exactly and we can work to find a solution to adding the 15 ilvls to the artifacts.

Kakjens commented:
Thanks for video. Seeing how it looks makes more understandable.
Was writing the code of branch ilvl_scantooltip while was half-asleep, so couldn't imagine how to merge the code for weapons with the code for other gear. Somehow can do it now.
If average item level returns the right values for upgraded artifacts, it might be simpler (and faster) to calculate artifact ilvl from it (especially if GetDetailedItemLevelInfo works for almost all items).

dejoblue commented:
Yeah, calculating it may be what we have to do for now. I want to get the relics for a feature addition later tho so I will explore it. 

I am going to work on integrating your code into this pull and should have it committed to this pull tonight/this morning sometime.

If you want to work on the math for artifact level that would be great. I will get to it or catch up with you on it after I get this pull cleaned up.

Kakjens commented:
Am sleepy, so try to do what you can.
About your [modified code](https://github.com/dejoblue/DejaCharacterStats/pull/112#issuecomment-330678110): I feel it won't work for either feral druids or prot warriors(more inclined to the latter). What should happen is that when setting text for main hand, the ilvl for offhand is not known, so it should stay 750 instead of 753.

dejoblue committed:
Updated Item Level Toltip Scan

This works for everything but Artifacts that have been upgraded via the
Netherlight Crucible.

dejoblue commented:
Video: https://youtu.be/MNE-vB58G3M

I have updated this pull with a new commit of what I consider the final version without the code to derive the ilvl of Artifacts upgraded via Netherlight Crucible. Look over it when you get a chance and make sure I don't have any extraneous or redundant thing, please and thank you.

In the video I discuss that I don't understand your concern for feral and prot and I show and explain my final code I have that just needs the math code to add the relic/Netherlight Crucible ilvl difference.

For right now:
We should do your math version idea to get this pushed to beta and released in the next week.
We can use the methods of tooltip scan and addition of items to derive the ilvl of Artifacts to add and derive the Bonus Experience stat in the future.

In the future:
I also want to add three slots above the weapon slots on the paper doll, with icon and mouseover display of the relics. That is why I was looking into getting the bonus ilvl from the relics and Artifact UI. But it is a global UI, there is not an API for the relics and it looks like it may be a mess. Once I get into it, if I can scrape the ilvl and it looks more efficient or easier or less code or makes more sense then we can change the method.

dejoblue committed:
Error fix

Moved local variables to only fire if an artifact item.

dejoblue commented:
Here is where I left off. still trying to figure out the logic of if there is not an offhand, like casters with a stave, then secondary is nil, etc. I will work more on it later today. If you have any ideas would be cool. May have to split them off after all like you did for the tooltip scan and make a function or two. Anyway here is the clusterfudge:

```
local function DCS_Item_Level_Center()
	local DCSDeriveArtILvl = 0
	local DCSDeriveArtILvlAve = 0
	for _, v in ipairs(DCSITEM_SLOT_FRAMES) do
		local itemLink = GetInventoryItemLink("player", v:GetID())
		if not itemLink then
			v.ilevel:SetFormattedText("")
		else
			local ITEM_LEVEL_PATTERN = ITEM_LEVEL:gsub("%%d", "(%%d+)")
			local tooltip = CreateFrame("GameTooltip", "iLevelScanTooltip", nil, "GameTooltipTemplate")
				tooltip:SetOwner(UIParent, "ANCHOR_NONE")
				tooltip:ClearLines()
				tooltip:SetHyperlink(itemLink)
			for i = 2, tooltip:NumLines() do
				local text = _G["iLevelScanTooltipTextLeft"..i]:GetText()
				if(text and text ~= "") then
					local value = tonumber(text:match(ITEM_LEVEL_PATTERN))
					if value then
						local _, _, itemRarity = GetItemInfo(itemLink)
						local r, g, b = GetItemQualityColor(itemRarity)
						v.ilevel:SetTextColor(r, g, b)
						DCSDeriveArtILvl = (DCSDeriveArtILvl + value) print(itemLink, value, DCSDeriveArtILvl) --Add all item level values
						-- if (CharacterSecondaryHandSlot.value == nil) then
						-- end
						if (itemRarity == 6) then 	--ALL Artifact items. If there were Artifact legs they would get the same ilvl text, since only weapons and we want them to be the same, this works and we don't care.
							local overall, equipped = GetAverageItemLevel() --print(equipped) --Get real average item level equipped
							local DCSMainHandSlot = GetDetailedItemLevelInfo(GetInventoryItemLink("player", (CharacterMainHandSlot:GetID())))
							local itemLink = GetInventoryItemLink("player", (CharacterSecondaryHandSlot:GetID()))
							if not itemLink then
								local DCSSecondaryHandSlot = DCSMainHandSlot
								print(DCSDeriveArtILvl)
								DCSDeriveArtILvl = (DCSDeriveArtILvl + DCSMainHandSlot) 
								print(DCSDeriveArtILvl)
							else
								local DCSSecondaryHandSlot = GetDetailedItemLevelInfo(itemLink)
							end					
							DCSArtDiffILvl = (((equipped) * 16) - DCSDeriveArtILvl) print(DCSArtDiffILvl, DCSDeriveArtILvl) --Calculate difference between total sum of real item level and calculated item level
							DCSArtILvl = (DCSArtDiffILvl/2) print(DCSArtILvl)
							--v.ilevel:SetText(DCSMainHandSlot + DCSArtILvl)
							print(DCSMainHandSlot + DCSArtILvl)
							if (DCSMainHandSlot > DCSSecondaryHandSlot) then
								v.ilevel:SetText(DCSMainHandSlot + DCSArtILvl)
							elseif (DCSSecondaryHandSlot > DCSMainHandSlot) then
								v.ilevel:SetText(DCSSecondaryHandSlot + DCSArtILvl)
							end
						else
							v.ilevel:SetText(value)
						end
						
					end
				end
			end
		end
	end
end
```

Kakjens commented:
My concerns were about your [previous](https://github.com/dejoblue/DejaCharacterStats/pull/112#issuecomment-330678110) code. The current pull should correctly set artifact ilvl.
I am now making test branch which you should like due to number of lines. Would like tests there. If the tests will be really bad, will write what exactly I don't like in your[ last code](https://github.com/dejoblue/DejaCharacterStats/pull/112#issuecomment-331138738).

Kakjens commented:
[Branch ](https://github.com/dejoblue/DejaCharacterStats/tree/artifact_trough_avg)for testing, especially on Crucible-upgraded artifacts (for example, your druid).

dejoblue committed:
Final Tiiltip Scan Update

Integrated Kakjens differential calculation code to derive Artifact
ilvl.

dejoblue commented:
Ahh, I see what you meant. I was coming from the other way. Cool. Works great. I will integrate it into this pull. we still have to use tooltip scan because heirlooms, warforged, etc., were off on my prot warrior.

dejoblue commented:
Integrated into this pull. Check it out see if there are any obvious mistakes.

There is still one issue, but it is a Blizzard issue. heirloom items are displaying the wrong ilvl. On the Heirloom Collection tab and also improper links. That is not something we can really do anything about other than put a Known Issue line in the description.

dejoblue committed:
Fixed Event Update Issues

Events fire before Artifact changes so we have to wait a fraction of a
second. Removed extraneous events registered.

dejoblue commented:
This should be ready to go once you give the go ahead, unless you find any other issues. I added two events, one for when you change talents specs and the other for when you change gear.

Bug Report: https://us.battle.net/forums/en/wow/topic/20759376668#1

Kakjens commented:
The more I thought about your bug report video, the less I understood. 
OK, so the first bug is that some heirloom erroneously scale up the displayed item level despite the heirloom not being upgraded to scale to the level of character. It happens on 1/3 upgraded. 2/3 heirlooms appear to be working correctly - the heirloom is not upgraded to scale past level 100. Mythic heirloom trinket is intended to scale to level 101 (and even beyond it). I feel that the tooltip itself in 1/3 cases is wrong, so we can't do nothing but wait for Blizzard to fix it. Indeed, a line about heirlooms not having the right item level is warranted.
The second bug is about GetDetailedLevelInfo returning wrong ilvl for old warforged/timewarped items. Tooltip scanning seems to be working in that case, so we have a workaround.
Intended to write about verification about GetDetailedLevelInfo not working correctly but saw it in the video.
Now changing this branch.

Kakjens committed:
Slight improvements of ITEM_LEVEL_PATTERN, tooltip scoping. Removal of r, g, b variables

Kakjens commented:
Those are the changes. After accepting merge.
